src/app/clusters/scenes-server/scenes-server.cpp

```diff
@@ -0,0 +1,993 @@
+ */
+
+#include "scenes-server.h"
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/scenes-server/SceneTableImpl.h>
+#include <app/reporting/reporting.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/error-mapping.h>
+#include <credentials/GroupDataProvider.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/Span.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/PlatformManager.h>
+
+using SceneTableEntry   = chip::scenes::DefaultSceneTableImpl::SceneTableEntry;
+using SceneStorageId    = chip::scenes::DefaultSceneTableImpl::SceneStorageId;
+using SceneData         = chip::scenes::DefaultSceneTableImpl::SceneData;
+using HandlerContext    = chip::app::CommandHandlerInterface::HandlerContext;
+using ExtensionFieldSet = chip::scenes::ExtensionFieldSet;
+using GroupDataProvider = chip::Credentials::GroupDataProvider;
+using SceneTable        = chip::scenes::SceneTable<chip::scenes::ExtensionFieldSetsImpl>;
+using AuthMode          = chip::Access::AuthMode;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Scenes {
+
+/// @brief Generate and add a response to a command handler context depending on an EmberAfStatus
+/// @tparam ResponseType Type of response, depends on the command
+/// @param ctx Command Handler context where to add reponse
+/// @param resp Response to add in ctx
+/// @param status Status to verify
+/// @return EmberAfStatus -> CHIP_ERROR
+template <typename ResponseType>
+CHIP_ERROR AddResponseOnError(CommandHandlerInterface::HandlerContext & ctx, ResponseType & resp, EmberAfStatus status)
+{
+    if (EMBER_ZCL_STATUS_SUCCESS != status)
+    {
+        resp.status = to_underlying(ToInteractionModelStatus(status));
+        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, resp);
+    }
+    return StatusIB(ToInteractionModelStatus(status)).ToChipError();
+}
+
+/// @brief Generate and add a response to a command handler context if err parameter is not CHIP_NO_ERROR
+/// @tparam ResponseType Type of response, depends on the command
+/// @param ctx Command Handler context where to add reponse
+/// @param resp Response to add in ctx
+/// @param status Status to verify
+/// @return CHIP_ERROR
+template <typename ResponseType>
+CHIP_ERROR AddResponseOnError(CommandHandlerInterface::HandlerContext & ctx, ResponseType & resp, CHIP_ERROR err)
+{
+    if (CHIP_NO_ERROR != err)
+    {
+        resp.status = to_underlying(StatusIB(err).mStatus);
+        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, resp);
+    }
+    return err;
+}
+
+ScenesServer ScenesServer::mInstance;
+
+ScenesServer & ScenesServer::Instance()
+{
+    return mInstance;
+}
+void ReportAttributeOnAllEndpoints(AttributeId attribute) {}
+
+CHIP_ERROR ScenesServer::Init()
```

Comment:
Why is this being called in the example app?  I would expect this to be called from either Server init if scenes is compiled in, or from `MatterScenesPluginServerInitCallback`...

Reply:
Putting this in the Server init will result in a circular dependency in gn builds as the scene server and table rely on the zap generated files and the zap generation step in the gn build has a dependency on the server. 

Putting this in the App was a way to avoid compiling scenes file if not needed without having to rely on ifdef. 

It was convenient to put this in the example App since that is where I was planning to register the scene handlers for various Cluster's and those were per App as users can choose what handlers to add, to use custom handlers, etc ( for example, what I am doing in this [branch](https://github.com/lpbeliveau-silabs/connectedhomeip/tree/feature/scenes-color-control-handler) )

It allowed for all scene initialization to happen in the same section of the code.

That being said it is totally possible that I missed a more appropriate mechanism to do so.

Reply:
Ah, interesting; I am surprised the scene handlers are app-specific.  I would have expected our built-in cluster impls to provide built-in handlers....

This is probably OK for now, but we should revisit the way this is initialized and used once more of the pieces are in place.  Please make sure we have an issue tracking it.

Reply:
Issue opened [here](https://github.com/project-chip/connectedhomeip/issues/26898).